### ISSUE TITLE
[Feature request] Temporarily deactivate when switching between workspaces

### ISSUE THREAD

**Issue by ghost** [LeonMatthes/mousefollowsfocus#4]
I had configured Dash to Panel to switch workspace on hover and scroll. With Mouse follow focus installed it became awkward that pointer location changes when trying to switch workspaces continuously.

So I suggest that you add:
- an inactive period on current workspace changes. when workspace change is detected, starts the timer and deactivates the extension functionality temporarily
- reactivation on timer ends, this certain amount of time should be roughly equals to the animation time
- if another workspace change is detected during this period, reset the timer (so that continuous switching works)
- If the duration could be made configurable through UI then it would be better

**Comment by thjderjktyrjkt** [comment]
Also, `focus_changed` event shouldn't be discarded during that period, because sometimes user may click a notification and it brings user to another workspace, in which case user may want mouse to follow the focus.

So the logic should be:
- Move the `if (actor)` part to a new `function`
- Create a timer. when receiving workspace changes event, start the timer
- When receiving focus changes event, check if the timer is running. If it does then just store the `actor`
- When receiving another focus changes event while there's already a pending `actor`, replace the existing one with the new `actor`, and reset the timer
- When timer is finished, trigger the new function with the stored `actor` where it does the real thing

**Comment by thjderjktyrjkt** [comment]
I don't know how to write javascript but this should showcase the logic:

```javascript
function do_real_thing(actor) {
    if (actor) {
        let rect = win.get_buffer_rect();

        let seat = Clutter.get_default_backend().get_default_seat();
        let [mouse_x, mouse_y, _] = global.get_pointer();

        if (cursor_within_window(mouse_x, mouse_y, win)) {
            dbg_log('pointer within window, discarding event');
        } else if (overview.visible) {
            dbg_log('overview visible, discarding event');
        } else {
            dbg_log('targeting new window');
            seat.warp_pointer(rect.x + rect.width / 2, rect.y + rect.height / 2);
        }

    }
}

let boolean sleeping = false;
let const actor_persistent = null;

function timer(time) {
    sleeping = true;
    dbg_log('deactivated temporarily for ' + time + 'ms due to workspace change');
    thread.sleep(time);
    sleeping = false;
    dbg_log('reactivated');
    do_real_thing(actor);
    dbg_log('sending the pending event');
    actor = null;
}


function workspace_changed() { //Trigger on workspace changes
    discard_existing_timer_thread_which_i_dont_know_how_to_do();
    thread.start(timer(time));
    dbg_log('workspace change event received');
}

function focus_changed(win) {
    const actor = get_window_actor(win);
    dbg_log('window focus event received');
    
    if(sleeping) {
        actor_persistent = actor;
        dbg_log('inhibited temporarily due to workspace change, storing actor');
    }
    else {
        do_real_thing(actor);
    }

}
```

**Comment by thjderjktyrjkt** [comment]
Yeah I know it won't work because i don't know how to write javascript

**Comment by thjderjktyrjkt** [comment]
oh wait i realized `win` should be used instead of `actor`, never mind

**Comment by thjderjktyrjkt** [comment]
and that `const` is really just constant value, not some special types, so it's guaranteed to not work.

**Comment by thjderjktyrjkt** [comment]
nope, can't understand it, it seems some gjs knowledge is required to write proper code for my FR to work, since either multethreading or time event is required

**Comment by ghost** [comment]
i've found this alternate extension which doesn't have the issue i had:
[Alt-Tab Move Mouse](https://github.com/buzztaiki/gnome-shell-extension-alt-tab-move-mouse)

**Comment by LeonMatthes** [comment]
I don't think disabling the extension for workspace switches is a road I want to go down.
Often times I switch workspaces using Super+1/Super+2 with Dash-To-Dock and in these cases I actually want my cursor to follow the window.

However, I have been playing with the idea of disabling the extension temporarily whenever mouse movement is detected.
That way, if you're on the keyboard, the extension will make your mouse follow the keyboard, but if you already have your hands on the mouse, it shouldn't jerk it around.
I've found this quite distracting, e.g. when interacting with popups.

This solution should solve your problem as well. The issue is I haven't yet found a way to get to the mouse events required for this.
If anyone has a pointer to documentation that can achieve this, that would be great.

**Comment by ghost** [comment]
Yes, that should resolve the issue for me as well.

Here's relevant documentation:

https://gjs-docs.gnome.org/clutter11~11_api/clutter.motionevent

Random things to take into consideration:
- If mouse only has tiny movement then just ignore it. Or even make better by making the threshold level configurable
- If mouse isn't moving but is clicking / dragging, then the mouse is also being held

---

Another problem I have is that cursor moves when top bar is clicked on and the maximized window is activated (because top bar is outside of the focused window). I don't know if this can be fixed by your strategy. [Alt-Tab Move Mouse](https://github.com/buzztaiki/gnome-shell-extension-alt-tab-move-mouse) doesn't have this issue though as it hooks into `Main.activateWindow` (only triggers when an existing window is activated by gnome shell). I've [made it](https://github.com/buzztaiki/gnome-shell-extension-alt-tab-move-mouse/pull/10) so newly created window also moves cursor but it's WIP implementation as I don't really know how to write gnome extension code.

**Comment by ghost** [comment]
Also your `get_focused_window` seems weird.

```javascript
function get_focused_window() {
    for (const actor of global.get_window_actors()) {
        if (actor.is_destroyed()) {
            continue;
        }

        const win = actor.get_meta_window();
        if (win.has_focus()) {
            return win;
        }
    }

    return undefined;
}
```

This would be the cleaner approach:

https://github.com/Dotrar/gnome-centre-focus/blob/main/extension.js#L50

```javascript
_moveMouse(){
    // get currently focused window 
    const win = global.display.focus_window;
    if (this._seat == null || win == null){
        return;
    }
    ...
}
```

**Comment by LeonMatthes** [comment]
> Also your `get_focused_window` seems weird.
> 
> ```js
> function get_focused_window() {
>     // ....
> }
> ```
> 
> This would be the cleaner approach:
> 
> https://github.com/Dotrar/gnome-centre-focus/blob/main/extension.js#L50
> 
> ```js
> _moveMouse(){
>     const win = global.display.focus_window;
>     ...
> }
> ```

Adapted your suggestion, thanks!
I did the loop over all actors as that was the first thing I found that worked.

I still find it hard to discover things in the Gnome JS API.
The Documentation on gjs-docs.gnome.org feels somewhat lacking 😕

**Comment by LeonMatthes** [comment]
> Yes, that should resolve the issue for me as well.
> 
> Here's relevant documentation:
> 
> https://gjs-docs.gnome.org/clutter11~11_api/clutter.motionevent
> 
> Random things to take into consideration:
> 
>     * If mouse only has tiny movement then just ignore it. Or even make better by making the threshold level configurable
> 
>     * If mouse isn't moving but is clicking / dragging, then the mouse is also being held
> 
> 
> Another problem I have is that cursor moves when top bar is clicked on and the maximized window is activated (because top bar is outside of the focused window). I don't know if this can be fixed by your strategy. [Alt-Tab Move Mouse](https://github.com/buzztaiki/gnome-shell-extension-alt-tab-move-mouse) doesn't have this issue though as it hooks into `Main.activateWindow` (only triggers when an existing window is activated by gnome shell). I've [made it](https://github.com/buzztaiki/gnome-shell-extension-alt-tab-move-mouse/pull/10) so newly created window also moves cursor but it's WIP implementation as I don't really know how to write gnome extension code.

RE MotionEvent:
That seems to be emitted by Clutter Actor. That means the extension would have to register itself with every actor in the scene.
Ideally I'd want a single attachment point for the extension, like the Clutter eventFilter.
Especially as we also definitely want to capture click events.
I have already tried registering an eventfilter with Clutter. Unfortunately that seems to miss events somewhat sporadically.
Can't tell why exactly.
Also, Wikipedia claims Clutter is deprecated since this year :thinking: 
So I don't really want to add a dependency for a core feature that may be abandoned in the near future.

The feature would however solve the topbar problem as well.
Unsure how to fix it otherwise :thinking: 
Just checking whether the mouse is in the topbar doesn't really help, as then if you alt-tab whilst the cursor is in the topbar, it wouldn't move.

I've also created a separate issue for this: #5
